File: src/main/java/com/github/caijh/commons/util/AesUtils.java
package com.github.caijh.commons.util;

import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.Key;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.KeyGenerator;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * Aes加密工具.
 */
public final class AesUtils {

    private static final String KEY_ALGORITHM = "AES";

    private static final String CIPHER_ALGORITHM = "AES/GCM/NoPadding";

    private AesUtils() {
        throw new AssertionError("no instance for you");
    }

    public static AesKey initAesKey() throws NoSuchAlgorithmException {
        KeyGenerator keyGenerator = KeyGenerator.getInstance(KEY_ALGORITHM);
        keyGenerator.init(256);
        SecretKey secretKey = keyGenerator.generateKey();
        byte[] iv = new byte[16];
        SecureRandom.getInstance("SHA1PRNG").nextBytes(iv);

        return new AesKey(secretKey.getEncoded(), iv);
    }

    private static Key toKey(byte[] key) {
        return new SecretKeySpec(key, KEY_ALGORITHM);
    }

    public static byte[] encrypt(byte[] data, AesKey key) throws NoSuchPaddingException, NoSuchAlgorithmException, InvalidKeyException, BadPaddingException, IllegalBlockSizeException, InvalidAlgorithmParameterException {
        Key k = toKey(key.getKey());
        Cipher cipher = Cipher.getInstance(CIPHER_ALGORITHM);
        GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(128, key.getIv());
        cipher.init(Cipher.ENCRYPT_MODE, k, gcmParameterSpec);
        return cipher.doFinal(data);
    }

    public static byte[] decrypt(byte[] data, AesKey aesKey) throws NoSuchPaddingException, NoSuchAlgorithmException, InvalidKeyException, BadPaddingException, IllegalBlockSizeException, InvalidAlgorithmParameterException {
        Key k = toKey(aesKey.getKey());
        Cipher cipher = Cipher.getInstance(CIPHER_ALGORITHM);
        GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(128, aesKey.getIv());
        cipher.init(Cipher.DECRYPT_MODE, k, gcmParameterSpec);
        return cipher.doFinal(data);
    }

    public static class AesKey {

        private final byte[] key;
        private final byte[] iv;

        public AesKey(final byte[] key, final byte[] iv) {
            this.key = key;
            this.iv = iv;
        }

        public byte[] getKey() {
            return key;
        }

        public byte[] getIv() {
            return iv;
        }

    }

}

File: src/main/kotlin/com/github/caijh/commons/util/AesUtils.kt
package com.github.caijh.commons.util

import java.security.*
import javax.crypto.*
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Aes加密工具.
 */
class AesUtils private constructor() {

    class AesKey(val key: ByteArray, val iv: ByteArray)

    companion object {
        private const val KEY_ALGORITHM = "AES"
        private const val CIPHER_ALGORITHM = "AES/GCM/NoPadding"

        @JvmStatic
        @Throws(NoSuchAlgorithmException::class)
        fun initAesKey(): AesKey {
            val keyGenerator = KeyGenerator.getInstance(KEY_ALGORITHM)
            keyGenerator.init(256)
            val secretKey = keyGenerator.generateKey()
            val iv = ByteArray(16)
            SecureRandom.getInstance("SHA1PRNG").nextBytes(iv)
            return AesKey(secretKey.encoded, iv)
        }

        private fun toKey(key: ByteArray): Key {
            return SecretKeySpec(key, KEY_ALGORITHM)
        }

        @JvmStatic
        @Throws(
            NoSuchPaddingException::class,
            NoSuchAlgorithmException::class,
            InvalidKeyException::class,
            BadPaddingException::class,
            IllegalBlockSizeException::class,
            InvalidAlgorithmParameterException::class
        )
        fun encrypt(data: ByteArray?, key: AesKey): ByteArray {
            val k = toKey(key.key)
            val cipher = Cipher.getInstance(CIPHER_ALGORITHM)
            val gcmParameterSpec = GCMParameterSpec(128, key.iv)
            cipher.init(Cipher.ENCRYPT_MODE, k, gcmParameterSpec)
            return cipher.doFinal(data)
        }

        @JvmStatic
        @Throws(
            NoSuchPaddingException::class,
            NoSuchAlgorithmException::class,
            InvalidKeyException::class,
            BadPaddingException::class,
            IllegalBlockSizeException::class,
            InvalidAlgorithmParameterException::class
        )
        fun decrypt(data: ByteArray?, aesKey: AesKey): ByteArray {
            val k = toKey(aesKey.key)
            val cipher = Cipher.getInstance(CIPHER_ALGORITHM)
            val gcmParameterSpec = GCMParameterSpec(128, aesKey.iv)
            cipher.init(Cipher.DECRYPT_MODE, k, gcmParameterSpec)
            return cipher.doFinal(data)
        }
    }

    init {
        throw AssertionError("no instance for you")
    }
}
